### PR TITLE
IPython integration - displaying blocks in Jupyter notebooks

### PR DESCRIPTION
**Pull request by shahinrostami**
Thank you so much for contributing to Datapane, please help us by filing out this PR template and ensure you have read the [CONTRIBUTING.md](../CONTRIBUTING.md).

It may take a few days to review your PR and merge it - please bear with us!

Thanks!

## Prerequsites

- [x] Has been tested locally
    - ✅ Tested with Jupyter Lab
    - ✅ Tested with Jupyter Notebook (classic)
    - ✅ Tested with VSCode
    - ✅ Tested exporting HTML file and opening locally
 
Notebook with some nice tests attached! It comes as a `zip` - GitHub didn't allow attaching an `ipynb`! 

[do_more_with_datapane.ipynb.zip](https://github.com/datapane/datapane/files/9685940/do_more_with_datapane.ipynb.zip)

## Proposed Changes

Improves the Datapane UX inside Jupyter notebooks, with proper IPython integration for blocks.

- `Stringify(LocalProcessor)` to support returning a Datapane app as a string.
- `ipython_template.html` to wrap the app in an iframe, with additional style/script for auto height calculation-adjustment.
- `App.stringify()` function to allow user to return an app as a string.
- IPython integration for `BaseElement`.

### CODE REVIEW

**Comment by shahinrostami** [comment]
Notebook with some nice tests attached! It comes as a `zip` - GitHub didn't allow attaching an `ipynb`! 

[do_more_with_datapane.ipynb.zip](https://github.com/datapane/datapane/files/9685901/do_more_with_datapane.ipynb.zip)

File: projects/python-client/src/datapane/client/api/report/processors.py [hunk 1]
```diff
@@ -502,6 +522,55 @@ def _open_server(host: str, port: int) -> None:
         webbrowser.open_new_tab(f"http://{host}:{port}")
 
 
+class Stringify(LocalProcessor):
+    """
+    Stringifies a given App as a single HTML string
+    """
+
+    served = False
+    template_name = "template.html"
```
Comment (mands) [inline projects/python-client/src/datapane/client/api/report/processors.py:531]: should `template_name` be a boolean, i.e. will there be other templates down the line?

```python
use_ipython_template: bool = False
```

File: projects/python-client/src/datapane/resources/local_report/ipython_template.html [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+    />
+    <script>
+      window.dpLocal = true;
+      {% if events %}
+        window.dpLocalViewEvent = {{events|tojson}};
+        window.dpReportId = "{{report_id}}";
+        window.dpAuthorId = "{{author_id}}";
+      {% endif %}
+    </script>
+    {% if standalone %}
```
Comment (mands) [inline projects/python-client/src/datapane/resources/local_report/ipython_template.html:25]: is the `standalone` boolean used when viewing via Jupyter? Do we want to hardcode `standalone` as False instead? i.e. we always require network access?

**Review by mands — APPROVED** [review]
LGTM - this is great - really looking forward to getting this in!

(I'm working on the Processor API a lot on the local apps branch, but should be simpler to use going forwards as well)

File: projects/python-client/src/datapane/client/api/report/processors.py (hunk 1 above)
Comment (shahinrostami) [reply projects/python-client/src/datapane/client/api/report/processors.py:531]: The plan is to allow for other templates down the line! In `processors.py` we are concerned with stringifying in general, as the feature now exists regardless of IPython integration!

**Pull request merged**